### PR TITLE
20220911-fixes-re-multi-test

### PR DESCRIPTION
**Pull request by douzzer**
wolfssl/wolfcrypt/sp_int.h and wolfcrypt/src/sp_int.c:

add `struct sp_int_minimal`, with same structure as `struct sp_int` but only one digit, to allow error-free access to `sp_int`s allocated with `ALLOC_SP_INT()` with fewer than `SP_INT_DIGITS` digits, and use the new type in `_sp_zero()` and `sp_init_size()` to eliminate `-Werror=array-bounds` on `_sp_zero()` under gcc-13.

tested with `wolfssl-mult-test.sh ... super-quick-check fips-140-3-dev-optest-acvp-sp-asm-gcc-latest linuxkm-mainline-pie-gcc-latest` at testing commit 7ccf1f2c3b9 (PR#346).

also adds support for `$KBUILD_EXTRA_FLAGS` to linuxkm/Makefile, needed to activate `linuxkm-mainline-pie-gcc-latest`.

### CODE REVIEW

**Comment by douzzer** [comment]
@SparkiDev note all tests have passed on this.  The "pending" on Options-A and FIPSv2 Regression are Jenkins->github integration glitches.

File: wolfcrypt/src/sp_int.c [hunk 1]
```diff
@@ -4394,10 +4394,20 @@ int sp_init(sp_int* a)
...
+    if (err == MP_OKAY) {
+    #ifdef HAVE_WOLF_BIGINT
+        wc_bigint_init(&a->raw);
+    #endif
+        _sp_zero(a);
+    }
 
     if (err == MP_OKAY) {
-        a->size = size;
+        ((sp_int_minimal *)a)->size = size;
```
Comment (SparkiDev) [inline wolfcrypt/src/sp_int.c:4410]: Why is this needed?
I would have thought from the description that it would only need to be where dp is accessed.

**Review by SparkiDev — CHANGES_REQUESTED** [review]

File: wolfcrypt/src/sp_int.c (hunk 1 above)
Comment (douzzer) [reply wolfcrypt/src/sp_int.c:4410]: It's needed because the compiler will warn, `-Warray-bounds`, on any dereference of a pointer to type `foo` if `foo` was allocated for less than `sizeof(foo)`.  The warning isn't driven by whether the accessed element is out of bounds, but by whether the object as a whole was adequately allocated.
```
In function ‘_sp_zero’,
    inlined from ‘sp_init’ at /home/wolfbot/tmp/wolfssl_test_workdir.34488/wolfssl/wolfcrypt/src/sp_int.c:4380:9,
    inlined from ‘sp_init_size’ at /home/wolfbot/tmp/wolfssl_test_workdir.34488/wolfssl/wolfcrypt/src/sp_int.c:4397:15,
b330196c28 (<sean@wolfssl.com> 2021-01-27 11:42:44 +1000 16085)             sp_init_size(b, 1);
    inlined from ‘sp_prime_is_prime’ at /home/wolfbot/tmp/wolfssl_test_workdir.34488/wolfssl/wolfcrypt/src/sp_int.c:16085:13:
539ef512fc (<sean@wolfssl.com> 2021-02-08 14:19:25 +1000 4355)     a->used = 0;
/home/wolfbot/tmp/wolfssl_test_workdir.34488/wolfssl/wolfcrypt/src/sp_int.c:4355:13: error: array subscript ‘sp_int[0]’ is partly outside array bounds of ‘unsigned char[24]’ [-Werror=array-bounds]
 4355 |     a->used = 0;
      |     ~~~~~~~~^~~
[...]
```

It's good that the compiler warns for this code pattern -- it is usually a mistake when an object is underallocated.

Adding `struct sp_int_minimal` allows for warning-free (and non-bug-prone) manipulation of (deliberately) underallocated `sp_int`s.

**Comment by SparkiDev** [comment]
Tried gcc-13 with -Werror=array-bounds with --enable-all and it didn't fail. (Not FIPS or kernel.)
What else do I have to do?

**Comment by douzzer** [comment]
So far I've only gotten it to trigger in the linuxkm build.  That's what the copy-paste is from.

My guess it's something to do with the options that Kbuild is passing to gcc.

Anyway the new `linuxkm-mainline-pie-gcc-latest` scenario in wolfssl-multi-test.sh exercises it, but only after this PR merges (the `$(KBUILD_EXTRA_FLAGS)` stuff in linuxkm/Makefile is needed).

File: wolfcrypt/src/sp_int.c [hunk 2]
```diff
@@ -4352,10 +4352,10 @@ static int _sp_mont_red(sp_int* a, sp_int* m, sp_int_digit mp);
  */
 static void _sp_zero(sp_int* a)
 {
-    a->used = 0;
-    a->dp[0] = 0;
+    ((sp_int_minimal *)a)->used = 0;
```
Comment (SparkiDev) [inline wolfcrypt/src/sp_int.c:4355]: Change this to: 
    sp_int_minimal* am = (sp_int_minimal*)a;

File: wolfcrypt/src/sp_int.c [hunk 3]
```diff
@@ -4394,10 +4394,20 @@ int sp_init(sp_int* a)
  */
 int sp_init_size(sp_int* a, int size)
 {
-    int err = sp_init(a);
+    int err = MP_OKAY;
+
+    if (a == NULL) {
```
Comment (SparkiDev) [inline wolfcrypt/src/sp_int.c:4399]: Pull common code out into _sp_init().
Have the size passed in.
I don't want to have to find everywhere that say HAVE_WOLF_BIGINT needs to be changed.

**Review by SparkiDev — CHANGES_REQUESTED** [review]

**Comment by SparkiDev** [comment]
What are the KBUILD_EXTRA_FLAGS?
grep found nothing so I assume it is something you add.
List them if possible.

**Comment by douzzer** [comment]
That's the recipe in `linuxkm-mainline-pie-gcc-latest`, referenced in the initial PR message.  That new scenario is already merged.  It is, in relevant part:
```
    KBUILD_EXTRA_FLAGS="CC=$GCC_LATEST HOSTCC=$GCC_LATEST" do_scenario_no_check "$scenario" CC="$GCC_LATEST" HOSTCC="$GCC_LATEST" "${ENABLE_ALL_TEST_FLAGS[@]}" --enable-aesni --enable-sp-asm --enable-linuxkm --enable-linuxkm-pie --enable-reproducible-build --enable-crypttests --with-linux-source="${linux[$linux_cur]}"
```
The only difference between the long-existing `linuxkm-mainline-pie` scenario and the new one is the `KBUILD_EXTRA_FLAGS=...` at the start, and the `CC="$GCC_LATEST" HOSTCC="$GCC_LATEST"` args to `configure`.

File: wolfcrypt/src/sp_int.c (hunk 2 above)
Comment (douzzer) [reply wolfcrypt/src/sp_int.c]: done

File: wolfcrypt/src/sp_int.c (hunk 3 above)
Comment (douzzer) [reply wolfcrypt/src/sp_int.c]: done

**Review by SparkiDev — APPROVED** [review]
